### PR TITLE
tests: Add fuzzing harness for Bech32 encoding/decoding

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harness for Bech32 encoding/decoding.

**Testing this PR**

Run:

```
$ make distclean
$ ./autogen.sh
$ CC=clang CXX=clang++ ./configure --enable-fuzz \
      --with-sanitizers=address,fuzzer,undefined
$ make
$ src/test/fuzz/bech32 -max_total_time=60
…
```

### CODE REVIEW

**Comment by jonatack** [comment]
When I try to test this, `make` fails with:
```
CXXLD    test/fuzz/address_deserialize
/usr/bin/ld: libbitcoin_util.a(libbitcoin_util_a-threadinterrupt.o): in function `UniqueLock<AnnotatedMixin<std::mutex>, std::unique_lock<std::mutex> >::~UniqueLock()':
/home/jon/projects/bitcoin/bitcoin/src/./sync.h:168: undefined reference to `LeaveCritical()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Full configure/make gist [here](https://gist.github.com/jonatack/691cb83400b4e84806d32ae36c656a62).

I have never been able to get the fuzzing to run when following the info in doc/fuzzing.md, so this may be unrelated to the PR. Any tips would be welcome.

**Comment by MarcoFalke** [comment]
Have you tried a `make distclean` before `./configure`?

**Comment by practicalswift** [comment]
@jonatack What @MarcoFalke said :) 

I've now updated OP with `make distclean; ./autogen.sh`. Will use that in the "how to test this PR template" in future OP:s for fuzzing PR:s.

```
$ make distclean
$ ./autogen.sh
$ CC=clang CXX=clang++ ./configure --enable-fuzz \
      --with-sanitizers=address,fuzzer,undefined
$ make
$ src/test/fuzz/bech32
…
```

**Comment by jonatack** [comment]
It looks like that solved it. Thanks!

How long does the test run -- until it's interrupted?
```
$ src/test/fuzz/bech32
INFO: Seed: 2164764235
INFO: Loaded 1 modules   (506453 inline 8-bit counters): 506453 [0x559733d74988, 0x559733df03dd), 
INFO: Loaded 1 PC tables (506453 PCs): 506453 [0x559733df03e0,0x5597345aa930), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 740 ft: 741 corp: 1/1b lim: 4 exec/s: 0 rss: 107Mb
#4	NEW    cov: 743 ft: 853 corp: 2/3b lim: 4 exec/s: 0 rss: 107Mb L: 2/2 MS: 2 ShuffleBytes-CopyPart-
...
#4005097	REDUCE cov: 829 ft: 2420 corp: 213/17Kb lim: 3381 exec/s: 2673 rss: 734Mb L: 44/641 MS: 3 CopyPart-EraseBytes-CopyPart-
#4132416	REDUCE cov: 829 ft: 2420 corp: 213/17Kb lim: 3502 exec/s: 2640 rss: 734Mb L: 7/641 MS: 4 ChangeBit-ChangeBit-ShuffleBytes-EraseBytes-
```

Concept ACK, initial code review looks good.

**Comment by practicalswift** [comment]
@jonasnick Yes, libFuzzer runs until interrupted :)  

Pass `-max_total_time=60` to run the fuzzer during 60 seconds, and `-help=1` to list all other libFuzzer options.

Recommended setup:

```
$ mkdir -p generated-test-corpus/
$ src/test/fuzz/bech32 -max_total_time=60 generated-test-corpus/
$ # now go into generated-test-corpus/ and investigate the generated test corpus :)
```

You might want to look at the libFuzzer documentation too if you haven't done that already :)

**Comment by jonasnick** [comment]
Thank you @practicalswift, I almost lost my mind trying to figure out libFuzzer today.

**Comment by practicalswift** [comment]
Sorry - wrong Jonas. Was meaning to mention @jonatack :)

File: test/fuzz/test_runner.py [hunk 1]
```diff
@@ -12,6 +12,10 @@
 import subprocess
 import logging
 
+# Fuzzers known to lack a seed corpus in https://github.com/bitcoin-core/qa-assets/tree/master/fuzz_seed_corpus
+FUZZERS_MISSING_CORPORA = [
+    "bech32",
+]
```
Comment (MarcoFalke) [inline test/fuzz/test_runner.py]: could remove?

**Comment by practicalswift** [comment]
@MarcoFalke Feedback addressed! Please re-review :)

**Review by jonatack — COMMENTED** [review]
ACK b7541705d0abfbddf682a0134f3fa8a8e1d06cdf

This fuzz test afaict verifies round-trip through bech32 Decode + Encode, comparing the initial random_string with the reencoded value, then Encode + Decode and asserts the correct values for the hrp (human-readable part as a string) and data.

Tested as per PR description, read code, ran tests.

```
((HEAD detached at origin/pr/17357))$ src/test/fuzz/bech32 -max_total_time=60
INFO: Seed: 3403448425
INFO: Loaded 1 modules   (506466 inline 8-bit counters): 506466 [0x5606fb3bc9e8, 0x5606fb43844a), 
INFO: Loaded 1 PC tables (506466 PCs): 506466 [0x5606fb438450,0x5606fbbf2a70), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 746 ft: 747 corp: 1/1b lim: 4 exec/s: 0 rss: 106Mb
.../...
#316157	DONE   cov: 835 ft: 2274 corp: 152/4114b lim: 122 exec/s: 5182 rss: 587Mb
###### Recommended dictionary. ######
"\xfd\xff\xff\xff" # Uses: 22198
"\x00\x00\x00\x00\x00\x00\x00\x00" # Uses: 3159
###### End of recommended dictionary. ######
Done 316157 runs in 61 second(s)

```

**Comment by jonatack** [comment]
I'm seeing appveyor failures:
```
bech32_tests.obj : error LNK2001: unresolved external symbol "bool __cdecl CaseInsensitiveEqual
```

**Comment by ryanofsky** [comment]
> I'm seeing appveyor failures:

Appveyor error is presumably caused by this PR and fixed by https://github.com/bitcoin/bitcoin/pull/17384#issuecomment-550305021

**Pull request merged**